cognite_toolkit/_cdf_tk/commands/dump_data.py

```diff
@@ -57,13 +69,108 @@ def _aggregate_count(self, hierarchies: list[str], data_sets: list[str]) -> int:
         raise NotImplementedError()
 
     @abstractmethod
-    def create_iterator(
-        self, hierarchies: list[str], data_sets: list[str], limit: int | None
-    ) -> Iterator[T_CogniteResourceList]:
+    def _get_resource_columns(self) -> list[SchemaColumn]:
+        """Get the columns for the schema."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def create_resource_iterator(self, limit: int | None) -> Iterable:
         raise NotImplementedError()
 
+    @abstractmethod
+    def _resource_processor(self, items: Iterable[T_CogniteResource]) -> list[tuple[str, list[dict[str, Any]]]]:
+        """Process the resources and return them in a format suitable for writing."""
+        raise NotImplementedError()
+
+    def _to_write(self, items: Iterable[T_CogniteResource]) -> list[dict[str, Any]]:
+        write_items: list[dict[str, Any]] = []
+        for item in items:
+            dumped = self.loader.dump_resource(item)
+            if "metadata" in dumped:
+                metadata = dumped.pop("metadata")
+                for key, value in metadata.items():
+                    dumped[f"metadata.{key}"] = value
+            if isinstance(dumped.get("labels"), list):
+                dumped["labels"] = [label["externalId"] for label in dumped["labels"]]
+                self._used_labels.update(dumped["labels"])
+            if "dataSetExternalId" in dumped:
+                self._used_data_sets.add(dumped["dataSetExternalId"])
+            write_items.append(dumped)
+        return write_items
+
+    def create_iterators(
+        self, format_: FileFormat, limit: int | None
+    ) -> Iterator[tuple[Schema, int, Iterable, Callable]]:
+        total = self.aggregate_count(tuple(self.hierarchies), tuple(self.data_sets))
+        columns = self._get_resource_columns()
+
+        iteration_count = total // 1000 + (1 if total % 1000 > 0 else 0)
+
+        yield (
+            Schema(
+                display_name=self.loader.display_name,
+                format_=format_,
+                columns=columns,
+                folder_name=self.loader.folder_name,
+                kind=self.loader.kind,
+            ),
+            iteration_count,
+            self.create_resource_iterator(limit),
+            self._resource_processor,
+        )
+        if self._used_data_sets:
+            yield self._data_sets()
+        if self._used_labels:
+            yield self._labels()
+
+    def _data_sets(self) -> tuple[Schema, int, Iterable, Callable]:
+        data_sets = self.client.data_sets.retrieve_multiple(
+            external_ids=list(self._used_data_sets), ignore_unknown_ids=True
+        )
+        loader = DataSetsLoader.create_loader(self.client)
+
+        def process_data_sets(items: DataSetList) -> list[tuple[str, list[dict[str, Any]]]]:
+            return [("", [loader.dump_resource(item) for item in items])]
```

Comment:
returning "" here seems a bit unusual? Can you add some comments here? A bit unclear to me